sql/src/main/java/io/crate/planner/consumer/MultiSourceGroupByConsumer.java

```diff
@@ -0,0 +1,231 @@
+            this.addPostAggregationProjection(plan, context, splitPoints, groupKeys, querySpec, outputs);
+
+            return plan;
+        };
+
+        /**
+         * Create and return the current execution plan.
+         */
+        private static Plan createExecutionPlan(ConsumerContext context, MultiSourceSelect multiSourceSelect) {
+            Plan plan = context.plannerContext().planSubRelation(multiSourceSelect, context);
+            boolean executesOnHandler = ExecutionPhases.executesOnHandler(
+                context.plannerContext().handlerNode(),
+                plan.resultDescription().nodeIds()
+            );
+            if (!executesOnHandler) {
+                // TODO: support distributed execution.
+                throw new UnsupportedFeatureException("Distribution execution is not supported");
+            }
+
+            return plan;
+        }
+
+        /**
+         * Adds the post aggregation protection to the given plan.
+         */
+        private static void addPostAggregationProjection(
+            Plan plan,
+            ConsumerContext context,
+            SplitPoints splitPoints,
+            List<Symbol> groupKeys,
+            QuerySpec querySpec,
+            List<Symbol> outputs
+        ) {
+            OrderBy orderBy = querySpec.orderBy().orElse(null);
+            Limits limits = context.plannerContext().getLimits(querySpec);
+            ArrayList list = new ArrayList<>(groupKeys);
+            list.addAll(splitPoints.aggregates());
+            Projection postAggregationProjection = ProjectionBuilder.topNOrEval(
+                list,
+                orderBy,
+                limits.offset(),
+                limits.finalLimit(),
+                outputs
+            );
+            plan.addProjection(postAggregationProjection, null, null, outputs.size(), null);
+        }
+
+        /**
+         * Adds the group protection to the given plan.
+         */
+        private void addGroupProtection(
+            Plan plan,
+            SplitPoints splitPoints,
+            List<Symbol> groupKeys
+        ) {
+            ProjectionBuilder projectionBuilder = new ProjectionBuilder(functions);
+
+            GroupProjection groupProjection = projectionBuilder.groupProjection(
+                splitPoints.leaves(),
+                groupKeys,
+                splitPoints.aggregates(),
+                AggregateMode.ITER_FINAL,
+                RowGranularity.CLUSTER
+            );
+            plan.addProjection(groupProjection, null, null, groupProjection.outputs().size(), null);
+        }
+
+        /**
+         * Adds the filter protection to the given plan.
+         */
+        private static void addFilterProtection(
+            Plan plan,
+            List<Symbol> groupKeys,
+            SplitPoints splitPoints,
+            QuerySpec querySpec
+        ) {
+            List<Symbol> collectOutputs = new ArrayList<>(
```

Comment:
instead of collectOutputs, aren't these the "postGrouping" outputs? with collect we usually refer to the data/output *before* the first projection.